### PR TITLE
Use the new REQUIRES_ACK flag and the EntryAckMessage

### PR DESCRIPTION
**Pull request by slinkydeveloper**
See https://github.com/restatedev/service-protocol/issues/53

### CODE REVIEW

File: src/types/types.ts [hunk 1]
```diff
@@ -53,19 +64,21 @@ class MessageType {
...
-  static hasRequiresAckFlag(messageTypeId: bigint): boolean {
-    return this.isCustom(messageTypeId);
+  static hasRequiresAckFlag(messageType: bigint): boolean {
+    return (
+      messageType !== START_MESSAGE_TYPE &&
+      messageType !== ERROR_MESSAGE_TYPE &&
+      messageType !== SUSPENSION_MESSAGE_TYPE &&
+      messageType !== ENTRY_ACK_MESSAGE_TYPE &&
+      messageType !== COMPLETION_MESSAGE_TYPE
+    );
```
Comment (tillrohrmann) [inline src/types/types.ts:74]: Can we say

```
messageType === SIDE_EFFECT_ENTRY_MESSAGE_TYPE
```

instead?

**Review by tillrohrmann — APPROVED** [review]
Thanks for adjusting the TS SDK wrt to the more explicit ack message. The changes look good to me. +1 for merging once we decide to merge https://github.com/restatedev/restate/pull/940.

File: src/types/types.ts (hunk 1 above)
Comment (slinkydeveloper) [reply src/types/types.ts:74]: no, the contract of this method is to return `true` if acks can be requested for this message, and this is true for any message that is an entry.

Comment (tillrohrmann) [reply src/types/types.ts:74]: True. Thanks for the clarification.

File: src/journal.ts [hunk 2]
```diff
@@ -189,14 +190,25 @@ export class Journal<I, O> {
...
       this.pendingJournalEntries.delete(m.entryIndex);
     } else {
-      if (journalEntry.messageType === p.SIDE_EFFECT_ENTRY_MESSAGE_TYPE) {
-        // Just needs and ack without completion
-        journalEntry.resolve(undefined);
-        this.pendingJournalEntries.delete(m.entryIndex);
-      } else {
-        //TODO completion message without a value/failure/empty and message is not a side effect
-      }
+      //TODO completion message without a value/failure/empty
```
Comment (slinkydeveloper) [inline src/journal.ts:193]: This one was just copied over from previous code.

File: src/journal.ts [hunk 3]
```diff
@@ -189,14 +190,25 @@ export class Journal<I, O> {
...
+
+  public handleEntryAckMessage(m: EntryAckMessage) {
+    // Get message at that entryIndex in pendingJournalEntries
+    const journalEntry = this.pendingJournalEntries.get(m.entryIndex);
+
+    if (
+      journalEntry === undefined ||
+      journalEntry.messageType !== p.SIDE_EFFECT_ENTRY_MESSAGE_TYPE
+    ) {
+      //TODO received completion message but for an entry that is either non existing, or is not a side effect entry
```
Comment (slinkydeveloper) [inline src/journal.ts:205]: Doing nothing here seems sane I guess, let me remove this todo

File: src/journal.ts (hunk 2 above)
Comment (slinkydeveloper) [reply src/journal.ts:193]: Looks like we should have an error here?

**Pull request merged**